module/zfs/arc.c

```diff
@@ -7856,9 +7980,11 @@ l2arc_write_eligible(uint64_t spa_guid, arc_buf_hdr_t *hdr)
 	 * 2. is already cached on the L2ARC.
 	 * 3. has an I/O in progress (it may be an incomplete read).
 	 * 4. is flagged not eligible (zfs property).
+	 * 5. is a prefetch and l2arc_noprefetch is set.
 	 */
 	if (hdr->b_spa != spa_guid || HDR_HAS_L2HDR(hdr) ||
-	    HDR_IO_IN_PROGRESS(hdr) || !HDR_L2CACHE(hdr))
+	    HDR_IO_IN_PROGRESS(hdr) || !HDR_L2CACHE(hdr) ||
+	    (l2arc_noprefetch && HDR_PREFETCH(hdr)))
```

Comment:
IIRC, this l2arc_noprefetch condition is actually handled elsewhere causing HDR_HAS_L2HDR() check to return false. No need to add it back here.

Reply:
I actually removed that line of code in `arc_read_done()`. As far as I can see, prefetches can be issued by bypassing `arc_read_done()` and in that case the flag HDR_L2CACHE will not be cleared.

So I moved that code to this location to be executed even if a prefetch bypasses `arc_read_done()`.